src/libclipper/include/clipper/task_executor.hpp

```diff
@@ -58,13 +58,90 @@ class PredictionCache {
   std::shared_ptr<metrics::RatioCounter> hit_ratio_;
 };
 
-template <typename Scheduler>
+struct DeadlineCompare {
+  bool operator()(const std::pair<Deadline, PredictTask>& lhs,
+                  const std::pair<Deadline, PredictTask>& rhs) {
+    return lhs.first > rhs.first;
+  }
+};
+
+// thread safe model queue
+class ModelQueue {
+ public:
+  ModelQueue() : queue_(ModelPQueue{}) {}
+
+  // Disallow copy and assign
+  ModelQueue(const ModelQueue&) = delete;
+  ModelQueue& operator=(const ModelQueue&) = delete;
+
+  ModelQueue(ModelQueue&&) = default;
+  ModelQueue& operator=(ModelQueue&&) = default;
+
+  ~ModelQueue() = default;
+
+  void add_task(PredictTask task) {
+    std::unique_lock<std::mutex> l(queue_mutex_);
+    Deadline deadline = std::chrono::high_resolution_clock::now() +
+        std::chrono::microseconds(task.latency_slo_micros_);
+    queue_.emplace(deadline, std::move(task));
+    queue_not_empty_.notify_all();
+  }
+
+  int get_size() {
+    std::unique_lock<std::mutex> l(queue_mutex_);
+    return queue_.size();
+  }
+
+  // TODO: If we implement the scheduler so that it calls
+  // get_earliest_deadline(), then uses that to compute the batch size,
+  // then calls dequeue with that batch size, it's possible for a new
+  // task with an earlier deadline to be submitted. I'm not sure what the
+  // correct behavior here should be.
+  Deadline get_earliest_deadline() {
+    std::unique_lock<std::mutex> l(queue_mutex_);
+    if (queue_.size() == 0) {
+      queue_not_empty_.wait(l, [this] { return queue_.size() > 0; });
```

Comment:
Based on our discussion about task starvation, this should be a non-blocking function right? Maybe it should return a `boost::option`?